File: src/main/java/gargoyle/ct/util/CTNumberUtil.java
package gargoyle.ct.util;

import org.jetbrains.annotations.NotNull;

import java.lang.reflect.InvocationTargetException;
import java.text.MessageFormat;

@SuppressWarnings("StaticMethodOnlyUsedInOneClass")
public final class CTNumberUtil {
    private static final String METHOD_VALUE_OF = "valueOf";
    private static final String MSG_CANNOT_CREATE_INSTANCE_OF_0 = "Cannot create instance of {0}";

    private CTNumberUtil() {
    }

    @NotNull
    @SuppressWarnings({"unchecked", "ChainOfInstanceofChecks"})
    public static <T extends Number> T fromInt(@NotNull Class<T> type, int value) {
        try {
            if (type == Integer.class) {
                return (T) Integer.valueOf(value);
            }
            if (type == Long.class) {
                return (T) Long.valueOf(value);
            }
            if (type == Double.class) {
                return (T) Double.valueOf(value);
            }
            if (type == Float.class) {
                return (T) Float.valueOf(value);
            }
            if (type == Byte.class) {
                return (T) Byte.valueOf((byte) value);
            }
            return (T) type.getMethod(METHOD_VALUE_OF, String.class).invoke(null, String.valueOf(value));
        } catch (@NotNull IllegalAccessException | InvocationTargetException | NoSuchMethodException ex) {
            throw new IllegalArgumentException(ex);
        }
    }

    @SuppressWarnings({"unchecked", "ChainOfInstanceofChecks"})
    public static <T> T getDefault(@NotNull Class<T> type) {
        if (type == Integer.class) {
            return (T) Integer.valueOf(0);
        }
        if (type == Long.class) {
            return (T) Long.valueOf(0);
        }
        if (type == Double.class) {
            return (T) Double.valueOf(0);
        }
        if (type == Float.class) {
            return (T) Float.valueOf(0);
        }
        if (type == Byte.class) {
            return (T) Byte.valueOf((byte) 0);
        }
        T def;
        try {
            def = Number.class.isAssignableFrom(type) ?
                    (T) type.getMethod(METHOD_VALUE_OF, String.class).invoke(null, "0") :
                    type.getConstructor().newInstance();
        } catch (@NotNull IllegalAccessException | InvocationTargetException | NoSuchMethodException | InstantiationException
                ex) {
            throw new IllegalArgumentException(MessageFormat.format(MSG_CANNOT_CREATE_INSTANCE_OF_0, type), ex);
        }
        return def;
    }

    public static Integer getInteger(Object value) {
        return Integer.valueOf(String.valueOf(value));
    }

    public static int toRange(int value, int min, int max) {
        return Math.max(min, Math.min(max, value));
    }

    public static Integer toRange(Integer value, Integer min, Integer max) {
        return Math.max(min, Math.min(max, value));
    }

    public static <T extends Number> int toRange(T min, T max, T value) {
        return Math.max(toInt(min), Math.min(toInt(max), toInt(value)));
    }

    public static <T extends Number> int toInt(T value) {
        return Integer.parseInt(String.valueOf(value));
    }
}

File: util/src/main/kotlin/gargoyle/ct/util/util/CTNumberUtil.kt
package gargoyle.ct.util.util

import kotlin.reflect.KClass
import kotlin.reflect.full.functions

object CTNumberUtil {
    private const val METHOD_VALUE_OF = "valueOf"
    private const val MSG_CANNOT_CREATE_INSTANCE_OF_0 = "Cannot create instance of {0}"

    @Suppress("UNCHECKED_CAST")
    fun <T : Number> fromInt(type: KClass<T>, value: Int): T {
        return when (type) {
            Int::class -> {
                return value as T
            }

            Long::class -> {
                return value.toLong() as T
            }

            Double::class -> {
                return value.toDouble() as T
            }

            Float::class -> {
                return value.toFloat() as T
            }

            Byte::class -> {
                value.toByte() as T
            }

            else -> type.functions.first { it.name == METHOD_VALUE_OF }.call(value.toString()) as T
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun <T : Any> getDefault(type: KClass<T>): T {
        when (type) {
            Int::class -> {
                return (0) as T
            }

            Long::class -> {
                return (0) as T
            }

            Double::class -> {
                return (0.0) as T
            }

            Float::class -> {
                return (0f) as T
            }

            Byte::class -> {
                return (0.toByte()) as T
            }

            else -> {
                return if ((type is Number)) type.functions.first { it.name == METHOD_VALUE_OF }.call("0") as T
                else type.constructors.first { it.parameters.isEmpty() }.call()
            }
        }
    }

    fun getInteger(value: Any): Int = (value.toString()).toInt()

    fun toRange(value: Int, min: Int, max: Int): Int = min.coerceAtLeast(max.coerceAtMost(value))

    fun <T : Number> toRange(min: T, max: T, value: T): Int =
        toInt(min).coerceAtLeast(toInt(max).coerceAtMost(toInt(value)))

    fun <T : Number> toInt(value: T): Int = value.toString().toInt()
}
